implementation/src/test/java/io/smallrye/mutiny/operators/UniCacheTest.java

```diff
@@ -273,4 +366,152 @@ public void onFailure(Throwable ignored) {
+        sub2.assertItem(0);
+        sub3.assertItem(1);
+        sub3.assertItem(1);
+    }
+
+    @Test
+    public void testThatFailureIsRenewedEvenBeforeCacheDurationExceeded() {
+        AtomicInteger subscriptionCount = new AtomicInteger(0);
+        Uni uni = Uni.createFrom()
+                .deferred(() -> Uni.createFrom().failure(new Exception("" + subscriptionCount.getAndIncrement())))
+                .cacheEvents().atMost(Duration.ofMinutes(1));
+
+        UniAssertSubscriber<Integer> sub1 = new UniAssertSubscriber<>();
+        UniAssertSubscriber<Integer> sub2 = new UniAssertSubscriber<>();
+        UniAssertSubscriber<Integer> sub3 = new UniAssertSubscriber<>();
+        UniAssertSubscriber<Integer> sub4 = new UniAssertSubscriber<>();
+
+        uni.subscribe().withSubscriber(sub1);
+        uni.subscribe().withSubscriber(sub2);
+        uni.subscribe().withSubscriber(sub3);
+        uni.subscribe().withSubscriber(sub4);
+
+        assertThat(subscriptionCount.get()).isEqualTo(4);
+        sub1.assertFailure(Exception.class, "0");
+        sub2.assertFailure(Exception.class, "1");
+        sub3.assertFailure(Exception.class, "2");
+        sub4.assertFailure(Exception.class, "3");
+    }
+
+    @Test
+    /*
+        TODO: This test is flaky: sometimes it ran into:
+        `java.lang.AssertionError: The uni didn't completed successfully: java.lang.IllegalStateException: Invalid transition, expected to be in the HAS_SUBSCRIPTION state but was in [1|2]`
+     */
+    public void testThatOnlyInvalidCacheEventsDoesNotRace() {
+        AtomicInteger cacheValidationCount = new AtomicInteger(0);
+        // events are always invalid, every test causes a resubscribe to upstream, since `onItem` is propagated to UniCache subscripte
+        BiPredicate<Integer, Throwable> validator = (i, f) -> false;
+        // multiple subscriptions may (and will) retrieve a single item from upstream after invalidation, so we cannot verify upstream subscription count
+        Uni<Integer> uni = Uni.createFrom().deferred(() -> Uni.createFrom().item(0))
+                .cacheEvents().whilst(validator);
+
+        // Every second subscription causes a re-subscription to the upstream
+        // This must not cause race conditions
+        List<UniAssertSubscriber<Integer>> results = Collections.synchronizedList(new LinkedList<>());
+        Executor parallel = Executors.newFixedThreadPool(10);
+        Runnable subscribe = () -> {
+            UniAssertSubscriber<Integer> subscriber = new UniAssertSubscriber<>(false);
+            uni.subscribe().withSubscriber(subscriber);
+            results.add(subscriber);
+        };
+        for (int i = 0; i < 1000; i++) {
+            parallel.execute(subscribe);
+        }
+
+        Uni.createFrom().item(0).onItem().delayIt().by(Duration.ofMillis(500)).await().indefinitely();
+
+        assertThat(results).hasSize(1000);
+        results.forEach(UniAssertSubscriber::assertCompletedSuccessfully);
```

Comment:
Build https://github.com/smallrye/smallrye-mutiny/pull/240/checks?check_run_id=966254111
failed with `java.lang.IllegalStateException: Invalid transition, expected to be in the HAS_SUBSCRIPTION state but was in 2` 
thrown from https://github.com/smallrye/smallrye-mutiny/blob/master/implementation/src/main/java/io/smallrye/mutiny/operators/UniSerializedSubscriber.java#L76

I didn't get it (erroneous state is `1` sometimes, too); may it be, that from time of check (`if (state.compareAndSet(HAS_SUBSCRIPTION, DONE))`) to throwing the IllegalArgumentException the state may be changed?